File: src/frontend/src/api/predict.ts
import constants from "@/utils/constants";
import { TaskState } from "@/types/tasks";
import * as datetime from "@/utils/datetime";

// backend.api.predict.PredictParams
export interface PredictParams {
    data: datetime.SimpleDateTimeStringStampedData;
    app_name: string;
    gt: datetime.SimpleDateTimeStringStampedData | null;
    model_exp_name: string;
}

// backend.model.PredictionError
export interface PredictionError {
    name: string;
    value: number;
    unit: string;
}

// backend.model.PredictResponse
interface PredictResponse {
    pred: datetime.SimpleDateTimeStringStampedData;
    errors: PredictionError[] | null;
}


export async function getPrediction(params: PredictParams): Promise<PredictResponse> {
    const resp = await fetch(`${constants.backendApiUrl}/predict`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(params),
    });
    const respObj = await resp.json();
    respObj.pred = datetime.isoDataToSimple(respObj.pred);
    return respObj as PredictResponse;
}

export type CeleryTaskId = string;
export type AsyncResponse = {
    task_id: CeleryTaskId
}
export async function getAsyncPrediction(params: PredictParams): Promise<AsyncResponse> {
    const resp = await fetch(`${constants.backendApiUrl}/predict/async`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(params),
    });
    return resp.json() as unknown as AsyncResponse;
}


type PredProgressMsg = {
    percentage: number,
}
export type PredProgressResponse = {
    state: TaskState
    msg: PredProgressMsg
}
export async function getAsyncPredictionProgress(task_id: CeleryTaskId): Promise<PredProgressResponse> {
    const resp = await fetch(`${constants.backendApiUrl}/predict/async/progress/${task_id}`);
    return resp.json() as unknown as PredProgressResponse;
}

export async function getAsyncPredictionResults(task_id: CeleryTaskId): Promise<PredictResponse> {
    const resp = await fetch(`${constants.backendApiUrl}/predict/async/results/${task_id}`)
    const respObj = await resp.json();
    respObj.pred = datetime.isoDataToSimple(respObj.pred);
    return respObj as PredictResponse;
}

File: src/backend/api/predict.py
from fastapi import APIRouter
from celery.result import AsyncResult
import pydantic
from pydantic import BaseModel

from ..pred import pred
from ..types.pred import PredictParams, PredictResponse
from ..types.tasks import CeleryTaskId, TaskState
from ..types.tasks.pred import PredProgressResponse, PredProgressMsg


router = APIRouter(prefix="/predict")


@router.post("/")
async def predict(predict_params: PredictParams) -> PredictResponse:
    # sync
    pred_res = pred(predict_params)
    assert isinstance(pred_res, PredictResponse)
    return pred_res


class AsyncResponse(BaseModel):
    task_id: CeleryTaskId


@router.post("/async")
async def async_predict(predict_params: PredictParams) -> AsyncResponse:
    pred_task = pred(predict_params, sync=False)
    assert isinstance(pred_task, AsyncResult)
    assert isinstance(pred_task.id, str)
    return AsyncResponse(task_id=pred_task.id)


@router.get("/async/progress/{task_id}")
async def predict_progress(task_id: str) -> PredProgressResponse:
    pred_task = AsyncResult(task_id)
    pred_state_enum = TaskState(pred_task.state.upper())

    try:
        msg = PredProgressMsg.model_validate_json(pred_task.info)
    except pydantic.ValidationError:
        msg = PredProgressMsg(percentage=100)
    
    return PredProgressResponse(
        state=pred_state_enum,
        msg=msg,
    )


@router.get("/async/results/{task_id}")
async def predict_results(task_id: str) -> PredictResponse:
    pred_task = AsyncResult(task_id)
    return PredictResponse.model_validate_json(pred_task.result)
